akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -193,173 +277,70 @@ private[akka] class AffinityPool(
-            case OptionVal.None    ⇒
-          }
-          spins += 1
-          if (spins > maxSpins) {
-            state = Yielding
-            yields = 0
-          }
-        case Yielding ⇒
-          yields += 1
-          if (yields > maxYields) {
-            state = Parking
-            parkPeriodNs = minParkPeriodNs
-          } else Thread.`yield`()
-        case Parking ⇒
-          LockSupport.parkNanos(parkPeriodNs)
-          parkPeriodNs = Math.min(parkPeriodNs << 1, maxParkPeriodNs)
-      }
-    }
-
-    def reset(): Unit = {
-      spins = 0
-      yields = 0
-      state = NotIdle
-    }
-
-  }
-
-  private final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
-
-  private final class ThreadPoolWorker(val q: BoundedTaskQueue, val idleStrategy: IdleStrategy) extends Runnable {
-
-    private sealed trait WorkerState
-    private case object NotStarted extends WorkerState
-    private case object InExecution extends WorkerState
-    private case object Idle extends WorkerState
-
-    val thread: Thread = tf.newThread(this)
-    @volatile private var workerState: WorkerState = NotStarted
+  private final class AffinityPoolWorker(val q: BoundedAffinityTaskQueue, val idleStrategy: IdleStrategy) extends Runnable {
+    final val thread: Thread = tf.newThread(this)
+    @volatile private[this] var workerState: WorkerState = NotStarted
 
     def startWorker(): Unit = {
       workerState = Idle
       thread.start()
     }
 
-    private def runCommand(command: Runnable) = {
+    private[this] final def runCommand(command: Runnable): Unit = {
       workerState = InExecution
       try
         command.run()
       finally
         workerState = Idle
     }
-
-    override def run(): Unit = {
-
+    override final def run(): Unit = {
       /**
-       * Determines whether the worker can keep running or not.
-       * In order to continue polling for tasks three conditions
-       * need to be satisfied:
-       *
-       * 1) pool state is less than Shutting down or queue
-       * is not empty (e.g pool state is ShuttingDown but there are still messages to process)
-       *
-       * 2) the thread backing up this worker has not been interrupted
-       *
-       * 3) We are not in ShutDown state (in which we should not be processing any enqueued tasks)
+       * We keep running as long as we are Running
+       * or we're ShuttingDown but we still have tasks to execute,
+       * and we're not interrupted.
        */
-      def shouldKeepRunning =
-        (poolState < ShuttingDown || !q.isEmpty) &&
-          !Thread.interrupted() &&
-          poolState != ShutDown
-
-      var abruptTermination = true
-      try {
-        while (shouldKeepRunning) {
+      @tailrec def runLoop(): Unit = {
+        val ps = poolState // Single volatile read
+        if (((ps == Running) || (ps == ShuttingDown && !q.isEmpty)) && !Thread.interrupted()) {
```

Comment:
@zaharidichev I think this conditional will be quicker to execute, please confirm :)